Libmesh update 20250304

roystgnr opened the pull request:
- Quadrature API changes.  A few virtual `QBase` methods have had
  changed signatures, and some methods have been deprecated in favor
  of newer versions, requiring `Elem` inputs rather than mere
  `ElemType` inputs, to enable support for elements like arbitrary
  polygons and polyhedra where a single class can take on different
  topologies at run time.
- `FE` and `FEInterface` API changes.  Some methods have been
  deprecated in favor of newer versions, either in `Elem` or taking an
  `Elem` rather than an `ElemType` argument, also for runtime-topology
  compatibility.
- Add APIs to control `System` name prefixing from C++, not just from
  the command line.  This will enable fixes for a MOOSE bug with 
  independent preconditioning of multiple nonlinear systems, affecting
  the optimization module.
- Added an `integrate_slits` option to `JumpErrorEstimator`-based
  estimators
- Added an `OverlapCoupling` ghosting functor, useful for integrating
  on slit meshes or between manifolds of overlapping meshes.
- Bug fix: `DistributedMesh::add_point` now operates correctly when
  given a node id that was recently deleted from the mesh but has not
  yet been cleaned from the internal node container. 
- Bug fix: `Prism6::build_side_ptr()` now sets `subdomain_id()`; this
  fixes a failure in adaptive mesh refinement of Prism6 elements with
  subdomain-restricted variables.
- Bug fix: methods which elevated mesh element order no longer
  renumber nodes and elements while doing so.  This fixes potential
  iterator invalidations in some use cases, and it's more efficient.
- Bug fix: when copying nodes and elements from a source mesh into a
  target `DistributedMesh` that is too small to use all its processors,
  the target mesh was left in an invalid state.  This fixes errors
  triggered by certain mesh stitching operations.
- Bug fix for range estimation bug in FParser optimizer
- Minor FE code refactoring for more simplicity and readability
- Minor optimization to some hash (and thus `unordered_map`) use cases
- Cleaned up and commented code with high-precision floating-point
  constants

@lindsayad was hoping to get one more libMesh PR squeezed into this, and we can wait for that before merging, but I want to get the API changes into CI right away and see what surprises may be in store.

roystgnr committed:
FEInterface -> FEMap::inverse_map

libMesh::FEInterface::inverse_map is now deprecated.

roystgnr committed:
FEInterface -> Elem::on_reference_element()

The former function is now deprecated as redundant

roystgnr committed:
Use non-deprecated libMesh n_dofs_at_node()

Refs #13921

roystgnr committed:
Newsletter entries for libMesh submodule update

roystgnr committed:
Update libMesh submodule

- Quadrature API changes.  A few virtual `QBase` methods have had
  changed signatures, and some methods have been deprecated in favor
  of newer versions, requiring `Elem` inputs rather than mere
  `ElemType` inputs, to enable support for elements like arbitrary
  polygons and polyhedra where a single class can take on different
  topologies at run time.
- `FE` and `FEInterface` API changes.  Some methods have been
  deprecated in favor of newer versions, either in `Elem` or taking an
  `Elem` rather than an `ElemType` argument, also for runtime-topology
  compatibility.
- Add APIs to control `System` name prefixing from C++, not just from
  the command line.  This will enable fixes for a MOOSE bug with
  independent preconditioning of multiple nonlinear systems, affecting
  the optimization module.
- Added an `integrate_slits` option to `JumpErrorEstimator`-based
  estimators
- Added an `OverlapCoupling` ghosting functor, useful for integrating
  on slit meshes or between manifolds of overlapping meshes.
- Bug fix: `DistributedMesh::add_point` now operates correctly when
  given a node id that was recently deleted from the mesh but has not
  yet been cleaned from the internal node container.
- Bug fix: `Prism6::build_side_ptr()` now sets `subdomain_id()`; this
  fixes a failure in adaptive mesh refinement of Prism6 elements with
  subdomain-restricted variables.
- Bug fix: methods which elevated mesh element order no longer
  renumber nodes and elements while doing so.  This fixes potential
  iterator invalidations in some use cases, and it's more efficient.
- Bug fix: when copying nodes and elements from a source mesh into a
  target `DistributedMesh` that is too small to use all its processors,
  the target mesh was left in an invalid state.  This fixes errors
  triggered by certain mesh stitching operations.
- Bug fix for range estimation bug in FParser optimizer
- Minor FE code refactoring for more simplicity and readability
- Minor optimization to some hash (and thus `unordered_map`) use cases
- Cleaned up and commented code with high-precision floating-point
  constants

Refs #0

roystgnr committed:
Update versioner templates

roystgnr committed:
Update versioner hashes

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit b30db56c835158b7f36397e4339c493a28592480
+Subproject commit c4791b4af32e3864d5160663d2c9e61989fe8d69
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2736920/), step Versioner verify on 94e8077 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status    | hash    | to hash   | version            | to version         |
|-----------|-----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE OK | dabc922 | c2aeeb2   | 2025.02.25=mpich_0 | 2025.03.06=mpich_0 |
| moose-dev | CHANGE OK | fcaa73c | ca0a369   | 2025.03.03=mpich   | 2025.03.06=mpich   |

roystgnr commented:
Yeah, those runtime `--disable-deprecated` failures are the sort of thing I was looking for.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2736915/), step Docs: sync website on 94e8077 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30042/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2736962/), step Generate coverage on 94e8077 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>2aabd6</th>
      <th colspan=4>#30042 94e807</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.26%</td>
      <td>85.26%</td>
      <td>-0.00%</td>
      <td>75.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108581</td>
      <td>108572</td>
      <td>-9</td>
      <td>12</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18769</td>
      <td>18770</td>
      <td>+1</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30042/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30042/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30042/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 75.00% is less than the suggested 90.0%



This comment will be updated on new commits.

roystgnr committed:
Make MaxQpsThread compatible with new libMesh

This is also a little faster, skipping an unnecessary test.

This is *broken* if users try to use a sufficiently higher quadrature
order on faces than on interiors for some reason, but in that use case
the existing code was broken too.  Fixing this without a performance
regression in typical use cases is going to require a new libMesh API.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2747551/), step Versioner verify on f18838d wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | dabc922 | c2aeeb2   | 2025.02.25 build 0 | 2025.03.06 build 0 |
| moose-dev | CHANGE   | fcaa73c | ca0a369   | 2025.03.03         | 2025.03.06         |

loganharbour commented:
Already?!

lindsayad commented:
We needs it

roystgnr committed:
Avoid deprecated n_shape_functions overload

We need n_dofs() with an actual Elem if we want to be able to handle
polyhedra where the result depends on more than just the ElemType

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2748443/), step Versioner verify on 076fa94 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | dabc922 | c2aeeb2   | 2025.02.25 build 0 | 2025.03.06 build 0 |
| moose-dev | CHANGE   | fcaa73c | ca0a369   | 2025.03.03         | 2025.03.06         |

roystgnr committed:
Avoid deprecated n_shape_functions() call

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2752594/), step Versioner verify on 30eca1f wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | dabc922 | c2aeeb2   | 2025.02.25 build 0 | 2025.03.06 build 0 |
| moose-dev | CHANGE   | fcaa73c | ca0a369   | 2025.03.03         | 2025.03.06         |

loganharbour commented:
ETA on this?

roystgnr committed:
Init quadrature with the full element

roystgnr committed:
Tell libMesh the deprecated init is deliberate

I'm not looking forward to this breaking when someone tries to get_xyz()
or something, but I can't justify adding even a little more expense in
an inner loop based on a hypothetical future possibility.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2755883/), step Versioner verify on 411ba14 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | dabc922 | c2aeeb2   | 2025.02.25 build 0 | 2025.03.06 build 0 |
| moose-dev | CHANGE   | fcaa73c | ca0a369   | 2025.03.03         | 2025.03.06         |

lindsayad commented:
> `PC failed due to FACTOR_OUTMEMORY` in the `navier_stokes distributed` test step.

sporadic failure

GiudGiud commented:
> sporadic failure

i ll try to patch this

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2757256/), step Versioner verify on 3a78465 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | dabc922 | c2aeeb2   | 2025.02.25 build 0 | 2025.03.06 build 0 |
| moose-dev | CHANGE   | fcaa73c | ca0a369   | 2025.03.03         | 2025.03.06         |

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2759037/), step Versioner verify on 945292b wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | dabc922 | c2aeeb2   | 2025.02.25 build 0 | 2025.03.06 build 0 |
| moose-dev | CHANGE   | fcaa73c | ca0a369   | 2025.03.03         | 2025.03.06         |

roystgnr committed:
Update deprecated quadrature init()

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2760807/), step Versioner verify on 3f4285c wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | dabc922 | c2aeeb2   | 2025.02.25 build 0 | 2025.03.06 build 0 |
| moose-dev | CHANGE   | fcaa73c | ca0a369   | 2025.03.03         | 2025.03.06         |

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2761918/), step Versioner verify on 077bbab wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status   | hash    | to hash   | version            | to version         |
|-----------|----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE   | dabc922 | c2aeeb2   | 2025.02.25 build 0 | 2025.03.06 build 0 |
| moose-dev | CHANGE   | fcaa73c | ca0a369   | 2025.03.03         | 2025.03.06         |

moosebuild commented:
Job [Modules parallel](https://civet.inl.gov/job/2761945/) on 077bbab : invalidated by @roystgnr

Kicking &quot;Modules parallel&quot; - &quot;test:finite_volume/ins/turbulence/lid-driven.lid_driven_turb_capped/result_physics:  Solve Did NOT Converge!&quot; from a &quot;distributed threads&quot; step

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2761960/), step Verify coverage on 077bbab wanted to post the following:


The following coverage requirement(s) failed:
- Failed to generate `chemical_reactions` coverage rate (required: 92.0%)
- Failed to generate `contact` coverage rate (required: 87.0%)
- Failed to generate `electromagnetics` coverage rate (required: 94.0%)
- Failed to generate `external_petsc_solver` coverage rate (required: 84.0%)
- Failed to generate `fluid_properties` coverage rate (required: 84.0%)
- Failed to generate `fsi` coverage rate (required: 85.0%)
- Failed to generate `functional_expansion_tools` coverage rate (required: 81.0%)
- Failed to generate `geochemistry` coverage rate (required: 96.0%)
- Failed to generate `heat_transfer` coverage rate (required: 87.0%)
- Failed to generate `level_set` coverage rate (required: 85.0%)
- Failed to generate `misc` coverage rate (required: 32.0%)
- Failed to generate `navier_stokes` coverage rate (required: 77.0%)
- Failed to generate `optimization` coverage rate (required: 86.0%)
- Failed to generate `peridynamics` coverage rate (required: 77.0%)
- Failed to generate `phase_field` coverage rate (required: 85.0%)
- Failed to generate `porous_flow` coverage rate (required: 95.0%)
- Failed to generate `ray_tracing` coverage rate (required: 93.0%)
- Failed to generate `rdg` coverage rate (required: 63.0%)
- Failed to generate `reactor` coverage rate (required: 90.0%)
- Failed to generate `scalar_transport` coverage rate (required: 81.0%)
- Failed to generate `solid_mechanics` coverage rate (required: 84.0%)
- Failed to generate `solid_properties` coverage rate (required: 83.0%)
- Failed to generate `stochastic_tools` coverage rate (required: 88.0%)
- Failed to generate `subchannel` coverage rate (required: 87.0%)
- Failed to generate `thermal_hydraulics` coverage rate (required: 88.0%)
- Failed to generate `xfem` coverage rate (required: 80.0%)

roystgnr commented:
The coverage errors here are the usual "Multiple base SHAs" issue

loganharbour commented on framework/src/geomsearch/FindContactPoint.C:
```diff
@@ -49,7 +50,7 @@ void
                  FEBase * fe_elem,
                  FEBase * fe_side,
-                 FEType & fe_side_type,
+                 FEType & /* fe_side_type */,
```
Being safe by keeping the argument?

roystgnr replied:
By keeping the argument in the function signature I'm trying to avoid API breakage further downstream, and trying not to add more trouble to this PR than I have to.  If this were libMesh (or at least if it was an API I'd written) then I'd make a new overload without the obviated argument, and make the existing overload a shim to the new one and mark it `libmesh_deprecated();` but for others' code I figured that's the kind of thing that might warrant a little discussion first, so I just made the minimal change to get everything `--disable-deprecated` compatible for now.

By keeping the argument name in `/* comments */` I'm just mindlessly obeying old habits.  We do that deliberately in libMesh code where e.g. there are arguments that one subclass needs but another doesn't, so now I guess I do it thoughtlessly any time I'm fixing unused-argument warnings.

loganharbour commented:
Let's leave the final merge to @milljm

milljm commented:
Minus next being busted for something @GiudGiud says he will fix (today: https://github.com/idaholab/moose/pull/28235), I am ready for this.

cticenhour commented on scripts/tests/versioner_hashes.yaml:
```diff
@@ -863,3 +863,34 @@ fd5f7bc651f0d6a8e2d23e5d4c2c2dadca58aa0b: #29978
   wasp:
     full_version: 2025.02.25_0
     hash: 46ea1c0
+9c3c8d833ce9c5d8fbb748dc6b870a075f4b2827:
```
Noted that this is missing a ticket number comment while rebasing #29984. Will add in a commit in that PR for completeness.

lindsayad replied:
This must be something missing in `--build-template`

cticenhour replied:
I thought the specific ticket number was something we should be adding manually after copy-pasting the `versioner.py --summary` output into `versioner_hashes.yaml`.

lindsayad replied:
You're right. I never scrolled that far in the wiki instructions 😆 I do think that this third step should also be somehow automated and included in `--build-template`
